wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java

```diff
@@ -0,0 +1,207 @@
+import static org.testng.Assert.assertTrue;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.wsplugins.model.PluginFQN;
+import org.eclipse.che.api.workspace.shared.Constants;
+import org.mockito.testng.MockitoTestNGListener;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+
+@Listeners(MockitoTestNGListener.class)
+public class PluginFQNParserTest {
+
+  private static final String PLUGIN_FORMAT = "%s/%s:%s";
+  private static final String PLUGIN_FORMAT_WITHOUT_REGISTRY = "%s:%s";
+
+  private PluginFQNParser parser;
+
+  @BeforeClass
+  public void setUp() throws Exception {
+    parser = new PluginFQNParser();
+  }
+
+  @Test
+  public void shouldReturnEmptyListWhenNoPluginsOrEditors() throws Exception {
+    Map<String, String> attributes = ImmutableMap.of("testProperty", "testValue");
+
+    Collection<PluginFQN> result = parser.parsePlugins(attributes);
+
+    assertTrue(
+        result.isEmpty(),
+        "PluginFQNParser should return empty list when attributes does not contain plugins or editors");
+  }
+
+  @Test(dataProvider = "validAttributesProvider")
+  public void shouldParseAllPluginsAndEditor(
+      String desc, List<PluginFQN> expected, Map<String, String> attributes) throws Exception {
+    Collection<PluginFQN> actual = parser.parsePlugins(attributes);
+    assertEqualsNoOrder(actual.toArray(), expected.toArray(), desc);
+  }
+
+  @Test(
+      dataProvider = "invalidAttributeStringProvider",
+      expectedExceptions = InfrastructureException.class)
+  public void shouldThrowExceptionWhenPluginStringIsInvalid(String plugin) throws Exception {
+    Map<String, String> attributes = createAttributes("", plugin);
+    parser.parsePlugins(attributes);
+  }
+
+  @Test(
+      dataProvider = "invalidAttributeStringProvider",
+      expectedExceptions = InfrastructureException.class)
+  public void shouldThrowExceptionWhenEditorStringIsInvalid(String editor) throws Exception {
+    Map<String, String> attributes = createAttributes(editor, "");
+
+    parser.parsePlugins(attributes);
+  }
+
+  @Test(
+      expectedExceptions = InfrastructureException.class,
+      expectedExceptionsMessageRegExp = "Multiple editors.*")
+  public void shouldThrowExceptionWhenMultipleEditorsDefined() throws Exception {
+    Map<String, String> attributes = createAttributes("editor1,editor2", "");
+
+    parser.parsePlugins(attributes);
+  }
+
+  @Test(
+      expectedExceptions = InfrastructureException.class,
+      expectedExceptionsMessageRegExp = "Invalid.*duplicated")
```

Comment:
The full expected message is `String.format("Invalid Che tooling plugins configuration: plugin %s:%s is duplicated", ...)` and I think it makes sense to check full message to make sure that message is formatted correctly (message contains `ID:VERSION` but not other values). I started to ask it after I saw that Che Server respond with errors where messages are wrong formatted, and it's pretty easy to make mistake here especially during refactoring.
WDYT?
It makes less sense to check the full message that is not formatted but still makes sense. It's like double fields for putting a password (password, confirm password). Also, if the message becomes formatted - then the author should change tests accordingly, otherwise it would be changed at all.

Reply:
Yeah it makes sense, I usually default to keeping the match as general as possible (only to ensure exception is being thrown for the right reason) but validating that plugin id and version are included makes a lot of sense.